### PR TITLE
runtime: don't use heap allocator in programs

### PR DESCRIPTION
**Pull request by ripatel-fd**

### CODE REVIEW

**Review by jumpsiegel — APPROVED** [review]

File: src/flamenco/runtime/context/fd_exec_txn_ctx.h [hunk 1]
```diff
@@ -49,6 +49,8 @@ struct __attribute__((aligned(8UL))) fd_exec_txn_ctx {
   fd_borrowed_account_t borrowed_accounts[128];          /* Array of borrowed accounts accessed by this transaction. */
   uchar                 unknown_accounts[128];           /* Array of boolean values to denote if an account is unknown */
   fd_txn_return_data_t  return_data;                     /* Data returned from `return_data` syscalls */
+  fd_clock_timestamp_vote_t clock_timestamps[64];        /* Array of clock timestamp side effects */
+  ulong                 clock_timestamp_cnt;             /* Number of clock timestamp side effects */
```
Comment (topointon-jump) [inline src/flamenco/runtime/context/fd_exec_txn_ctx.h:53]:
```suggestion
  ulong                    clock_timestamps_cnt;         /* Number of clock timestamp side effects */
```

File: src/flamenco/runtime/program/fd_stake_program.c [hunk 2]
```diff
@@ -1308,6 +1324,8 @@ delegate( fd_exec_instr_ctx_t const *   ctx,
           fd_pubkey_t const *           signers[static FD_TXN_SIG_MAX] ) {
   int rc;
 
+  fd_valloc_t scratch_valloc = fd_scratch_virtual();
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_stake_program.c:1327]: Should we wrap the whole instruction processor in a `FD_SCRATCH_SCOPE`? To make the semantics clear?

File: src/flamenco/runtime/context/fd_exec_txn_ctx.h [hunk 3]
```diff
@@ -49,6 +49,8 @@ struct __attribute__((aligned(8UL))) fd_exec_txn_ctx {
   fd_borrowed_account_t borrowed_accounts[128];          /* Array of borrowed accounts accessed by this transaction. */
   uchar                 unknown_accounts[128];           /* Array of boolean values to denote if an account is unknown */
   fd_txn_return_data_t  return_data;                     /* Data returned from `return_data` syscalls */
+  fd_clock_timestamp_vote_t clock_timestamps[64];        /* Array of clock timestamp side effects */
```
Comment (topointon-jump) [inline src/flamenco/runtime/context/fd_exec_txn_ctx.h:52]: Could you document why 64 is appropiate?

File: src/flamenco/runtime/program/fd_vote_program.c [hunk 4]
```diff
@@ -488,63 +492,69 @@ is_uninitialized( fd_vote_state_versioned_t * self ) {
...
+        .is_empty = 1,
+      },
+      .epoch_credits  = state->epoch_credits,   /* move */
+      .last_timestamp = state->last_timestamp,  /* deep copy */
     };
-    memset( current.prior_voters.buf, 0, sizeof( current.prior_voters.buf ) );
-    current.epoch_credits  = state.epoch_credits;
-    current.last_timestamp = state.last_timestamp;
+
+    /* Move objects */
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_vote_program.c:521]: Just a thought: maybe we should have an `fd_move` which sets the old location to `NULL`?

**Review by topointon-jump — APPROVED** [review]

**Review by topointon-jump — APPROVED** [review]

File: src/flamenco/runtime/context/fd_exec_txn_ctx.h (hunk 3 above)
Comment (lheeger-jump) [inline src/flamenco/runtime/context/fd_exec_txn_ctx.h:52]: I am not immediately sure where it is but where are the clock timestamps being recorded so the clock can use them after the txn is done?

**Pull request merged**